File: snobot_sim_java/java/snobot_sim_java/com/snobot/simulator/motor_sim/SimpleMotorSimulator.java
package com.snobot.simulator.motor_sim;

public class SimpleMotorSimulator implements IMotorSimulator
{

    private SimpleMotorSimulationConfig mConfig;

    private double mVoltagePercent;
    private double mVelocity;
    private double mPosition;

    public SimpleMotorSimulator(SimpleMotorSimulationConfig aConfig)
    {
        mConfig = aConfig;
    }

    @Override
    public void setVoltagePercentage(double speed)
    {
        mVoltagePercent = speed;
    }

    @Override
    public double getVoltagePercentage()
    {
        return mVoltagePercent;
    }

    @Override
    public double getAcceleration()
    {
        return 0;
    }

    @Override
    public double getVelocity()
    {
        return mVelocity;
    }

    @Override
    public double getPosition()
    {
        return mPosition;
    }

    @Override
    public double getCurrent()
    {
        return 0;
    }

    @Override
    public void reset()
    {
        reset(0, 0, 0);
    }

    @Override
    public void reset(double aPosition, double aVelocity, double aCurrent)
    {
        mPosition = aPosition;
        mVelocity = aVelocity;
    }

    @Override
    public void update(double aUpdateTime)
    {
        mVelocity = mConfig.mMaxSpeed * mVoltagePercent;
        mPosition += mVelocity * aUpdateTime;
    }

    public SimpleMotorSimulationConfig getConfig()
    {
        return mConfig;
    }
}

File: snobot_sim/src/main/native/cpp/SnobotSim/MotorSim/SimpleMotorSimulator.cpp
/*
 * SimpleMotorSimulator.cpp
 *
 *  Created on: May 5, 2017
 *      Author: PJ
 */

#include "SnobotSim/MotorSim/SimpleMotorSimulator.h"

SimpleMotorSimulator::SimpleMotorSimulator(double aMaxSpeed) :
        mMaxSpeed(aMaxSpeed),
        mVoltagePercent(0),
        mVelocity(0),
        mPosition(0)
{
}

SimpleMotorSimulator::~SimpleMotorSimulator()
{
}

std::string SimpleMotorSimulator::GetSimulatorType()
{
    return GetType();
}

double SimpleMotorSimulator::GetMaxSpeed()
{
    return mMaxSpeed;
}

void SimpleMotorSimulator::SetVoltagePercentage(double aSpeed)
{
    mVoltagePercent = aSpeed;
}

double SimpleMotorSimulator::GetVoltagePercentage()
{
    return mVoltagePercent;
}

double SimpleMotorSimulator::GetAcceleration()
{
    return 0;
}

double SimpleMotorSimulator::GetCurrent()
{
    return 0;
}

double SimpleMotorSimulator::GetVelocity()
{
    return mVelocity;
}

double SimpleMotorSimulator::GetPosition()
{
    return mPosition;
}

void SimpleMotorSimulator::Reset()
{
    Reset(0, 0, 0);
}

void SimpleMotorSimulator::Reset(double aPosition, double aVelocity, double aCurrent)
{
    mPosition = aPosition;
    mVelocity = aVelocity;
}

void SimpleMotorSimulator::Update(double aUpdateTime)
{
    mVelocity = mMaxSpeed * mVoltagePercent;
    mPosition += mVelocity * aUpdateTime;
}
